docs/design/proposals/multi-version/multi-version.md

```diff
@@ -0,0 +1,259 @@
+at various stages and levels. This makes it challenging to implement
+webhooks that can handle multiple versions and deprecations effectively.
+
+## Overview of the design
+
+The proposed design aims to address the challenges of multi-versioning
+and generating mutating webhooks in Amazon Controllers for Kubernetes (ACK).
+It involves modifications to the ACK/code-generator tool and the implementation
+of a webhook server to handle multiple versions and deprecations of Custom
+Resource Definitions (CRDs). The solution uses a combination of Kubernetes API
+machinery and controller-runtime libraries, and the document outlines a
+high-level plan for implementation, including updating the code-generator
+tool and modifying ACK's controllers. Additionally, the document discusses
+the testing strategies that will be employed to ensure backward compatibility.
+
+### Rethinking the ACK versioning
+
+Before jumping to any special topic, we need to answer two important questions:
+- Do we still want to bump all the CRD versions all at once? meaning that
+  if we rename a field in CRD A, CRD B C D and E will also need a version bump
+  (and a mutating webhook handler) even if they are not modified?
+- How do we deprecate/remove versions in a programatical way?
+
+#### CRDs Mono Version vs Group Version
+
+The answer to the questions above will impact the design of [`metadata.yaml`]
+[iam-metadata.yaml]. And how we think about ACK CRD versioning and how
+version deprecation happens. Do we deprecate a set of CRD versions? or only
+one specific CRD version? unfortunetaly picking one of those solutions is kinda
+a hard one way door solution to say the least.
+
+CRD Group versioning:
+
+- Pros:
+1. Current ACK code-generator works this way
+2. Easier code generation (import paths, packaging etc...)
+3. Easier CRD Version difference calculation
+4. Doesn't enormously impact the rest of the features (references etc...)
+5. No need to change `metadata.yaml`
+
+- Cons:
+1. Confusing for users
+2. Little CRD Changes means we will bump all the CRDs versions
+3. Not very kubernetes style.
+
+CRD Mono versioning:
+
+- Pros:
+1. Similar to k8s style
+2. Easier deployments (risk reduced when upgrading CRDs)
+3. Less error zone as we will target resources one by one
+
+- Cons:
+1. Very tricky in code generation
+2. Needs a rethinking of generator.yaml + metadata.yaml
+3. We don't know the impact on existing features (references etc...)
+
+#### CRDs version deprecation
+
+Curently we only have `metadata.yaml` as a tool to deprecate versions, see
+APIInfoStatuses contants in [code-generator/pkg/metadata][code-gen/pkg/md]. Which
+works very well with `pkg/model/multiversion` package. In case we want to go with
+CRD Mono Version way, we will have to change this model to be Resource specific
+rather than version specific.
+
+### Webhook server
+
+Controller runtime already offers webhook server setup as a builtin mechanisms, and
```

Comment:
"Controller runtime" -> "Kubernetes controller-runtime" (provide a link to relevant doc?)
"a builtin mechanisms" -> "a built-in mechanism"